### DISCUSSION TITLE
UserLoginFailure: Transaction not successfully started thrown in multiple tests

### DISCUSSION THREAD

**Discussion by sguilhen** [keycloak/keycloak#11176] [Archive]
I've finished the implementation of the user login failures JPA storage and while testing it I came across this error in tests that remove more than one user via `UserManager`:

```
java.lang.IllegalStateException: Transaction not successfully started
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:98)
	at org.keycloak.models.map.storage.jpa.JpaMapKeycloakTransaction.commit(JpaMapKeycloakTransaction.java:176)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:146)
```

[UserSessionProviderTest](https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderTest.java#L88) is a good example of a test that causes this error. Basically this is what happens:

1. Test calls `UserManager.removeUser` to remove more than one user as part of the same transaction.
2. For each user that is removed:
    - a `UserRemovedEvent` is generated and is handled by the [MapUserLoginFailureProviderFactory](https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/loginFailure/MapUserLoginFailureProviderFactory.java#L66).
   - the factory creates a new `MapUserLoginFailureProvider`, which on creation calls `store.createTransaction(session)`. 
   - the code for the JPA store checks the session for an existing transaction using a predefined key `jpa-map-tx-modelType.hashCode`, returning the previous `tx` if one is found in the session or creating a new `tx` otherwise.
   - the user login failure provider enlists the tx.

So when removing the first user, a new transaction is created for the user login failure and enlisted. From that point on, every time the `MapUserLoginFailureProviderFactory` handles the removal event, it creates a `MapUserLoginFailureProvider` that in turn fetches the previously created `tx` from the session and enlists it again.

When the overall transaction is committed, the user login failure transaction is committed multiple times because it has been enlisted more than once while handling the user removal event, which causes the exception described above.

I could think of a few ways to get around it:

1. `removeUser` is called multiple times in the same `UserManager` only by the tests. Our endpoints offer no bulk user removal, so it always processes one removal per transaction. That means we can probably adjust the tests to, for example, remove the created users calling the endpoint directly for each user (`adminClient.realm("test").users().get(userId1).remove();`), but this requires keeping track o the ids of the users created in each test.
2. make all the `enlist` methods in `DefaultKeycloakTransactionManager` check if the same instance has already been enlisted - this would prevent multiple `commit()` calls to the same transaction object.
3. Adjust the `MapUserLoginFailureProvider`'s constructor so that it somehow checks if the `tx` already exists in the session before calling `enlistAfterComplete`. This would ensure that only the provider that created the transaction that is currently active in the session gets to enlist it.

I'm currently leaning towards option 2 and make the `enlist` methods check if the list of transactions already contain the transaction being enlisted to avoid enlisting it twice. Option 1 works but will require many changes to all the tests that remove users. Option 3 is feasible although I would need to investigate a viable option for looking up the session because the key is constructed by the `JpaMapStorageProvider` and is not available to the user login failure provider.

**Comment by ahus1** [comment]
Hi @sguilhen, looking at the code option 2 seems to be the smallest change. 

Looking a bit more closer at the implementation I did for LDAP, it is strange that the code in the storage provider already decides if a new transaction should be created or an existing should be used, and some other code would decide if it should be enlisted. Also, that other code would decide if it would be enlist() or enlistAfterCompletion(). 

```
            MapKeycloakTransaction<V, M> sessionTx = session.getAttribute(sessionTxPrefix + modelType.hashCode(), MapKeycloakTransaction.class);
            if (sessionTx == null) {
                sessionTx = factory.createTransaction(session, modelType);
                session.setAttribute(sessionTxPrefix + modelType.hashCode(), sessionTx);
            }
            return sessionTx;
``` 

Therefore I would propose the following change: the one who creates the transaction should also enlist it with the session, in this case the one who calls factory.createTransaction() would then also enlist it with the session.

WDYT?

**Reply by sguilhen** [reply]
Yeah, I agree with the proposal. The challenge here is that each `MapStorageProvider.createTransaction` impl is different. `HotRodMapStorage` doesn't use the session, while `JpaMapStorageProvider`, `LapMapStorageProvider` and `ConcurrentHashMapStorage` all use the session but with different keys. So the particular `Map*Provider` that is creating and enlisting the transaction has no clue about what format it should use to look up the session.

**Reply by sguilhen** [reply]
One idea would be to add a method to the `MapStorage` that could be used to query existing tx. Something like `getCurrentTransaction(KeycloakSession session)`, which would be implemented by stores to return the value that is currently found in the session. Then, if the returned tx is null, the code calls `createTransaction` followed by `enlist` as it does today. So the logic would be something like

```
// constructor
public Map*Provider(KeycloakSession session, MapStorage<Map*Entity, *Model> realmStore) {
    this.session = session;
    this.tx = realmStore.getCurrentTransaction(session);
    if (this.tx == null) {
        this.tx = realmStore.createTransaction(session);
        session.getTransactionManager().enlist(this.tx); // or enlistAfterCompletion
    }
}
```

**Reply by ahus1** [reply]
I had a look at HotRodMapStorage and this is the current code:

```
    @Override
    public MapKeycloakTransaction<V, M> createTransaction(KeycloakSession session) {
        Map<SearchableModelField<? super M>, MapModelCriteriaBuilder.UpdatePredicatesFunc<K, V, M>> fieldPredicates = MapFieldPredicates.getPredicates((Class<M>) storedEntityDescriptor.getModelTypeClass());
        return new ConcurrentHashMapKeycloakTransaction<>(this, keyConverter, cloner, fieldPredicates);
    }
``` 

It can be changed to: 

```
    @Override
    public MapKeycloakTransaction<V, M> createTransaction(KeycloakSession session) {
        Map<SearchableModelField<? super M>, MapModelCriteriaBuilder.UpdatePredicatesFunc<K, V, M>> fieldPredicates = MapFieldPredicates.getPredicates((Class<M>) storedEntityDescriptor.getModelTypeClass());
        ConcurrentHashMapKeycloakTransaction<K, V, M> transaction = new ConcurrentHashMapKeycloakTransaction<>(this, keyConverter, cloner, fieldPredicates);
        session.getTransactionManager().enlist(transaction); // or enlistAfterCompletion
        return transaction;
    }
``` 

I assume you don't want to do this as you want to hide / avoid duplication of `session.getTransactionManager().enlist(transaction); // or enlistAfterCompletion` in the different implementations. 

At the same time it feels strange that all the ones that caller of createTransaction() would also enlist the transaction. 

Looking at both sides of the argument, I'd rather not have another method and put the enlisting into createTransaction() as a transaction that is not enlisted wouldn't be of much use. 

If the createTransaction() wouldn't know about enlist vs. enlistAfterCompletion, that could be provided as a callback passed as a parameter (although that makes it less appealing).

**Reply by ahus1** [reply]
Some afterthoughts after talking to @sguilhen: If the decision of enlist vs. enlistAfterCompletion stays with the caller, a transaction could be enlisted within both or - depending on a future logic - with one or the other depending on who is calling it first.

**Reply by sguilhen** [reply]
@hmlnarik do you have any thoughts on how to proceed with this? I was able to get the user login failures JPA store working with all tests by adding checks to the transaction manager so it doesn't enlist the same instance, but it feels like we could do better so that this situation doesn't even happen in the first place.

**Reply by hmlnarik** [reply]
> Yeah, I agree with the proposal. The challenge here is that each `MapStorageProvider.createTransaction` impl is different. `HotRodMapStorage` doesn't use the session, while `JpaMapStorageProvider`, `LapMapStorageProvider` and `ConcurrentHashMapStorage` all use the session but with different keys. So the particular `Map*Provider` that is creating and enlisting the transaction has no clue about what format it should use to look up the session.

Note that there might be two separate transactions active at the same time, say for a configuration where there are two LDAP connections targetting separate servers configured in the tree store, both active at the same time.

So it is not desirable for a storage instance to interfere with a transaction created with the other instance. Particular `Map*Provider` should not be looking up the transaction, it should be the storage that owns and understands how to look the transaction up, e.g. https://github.com/keycloak/keycloak/blob/fb81242658e86e9ae1fc43169878ec430210d61c/model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java#L131-L132

> Some afterthoughts after talking to @sguilhen: If the decision of enlist vs. enlistAfterCompletion stays with the caller, a transaction could be enlisted within both or - depending on a future logic - with one or the other depending on who is calling it first.

Where the transaction is enlisted depends on the semantics of the caller and thus cannot be enlisted at both, nor arbitrarily altered in FCFS manner.



> [...] it feels like we could do better so that this situation doesn't even happen in the first place.

I agree. The legacy JPA transaction is never enlisted twice since it is created in a separate provider. Sadly, we don't have this comfort here, as there might be several JPA servers being targetted (cf. several LDAP servers example above).

It seems that once we solve the question how to get unique key to a session attribute, we'd be done. In case of JPA, the transaction should be the same for the same connection which in turn is represented by a respective map storage `ProviderFactory`. IOW, contrary to `ConcurrentHashMapStorage` above, rather than hash code on that storage, we need an identifier that is passed to each map storage from the factory and used as the key (could be a hash code, but unique a identifier obtained from an `AtomicInteger` could be even better as it is collision-free). WDYT?

**Reply by sguilhen** [reply]
That seems reasonable to me. So each storage instance is created with a different identifier that it will use in it's `createTransaction` method, meaning that you will only get to reuse the same tx if the same storage instance is being called to create transactions multiple times?

**Reply by hmlnarik** [reply]
Almost - this works in case of CHM or HotRod where the transaction is processed per each storage (i.e. per each model class). In JPA the transaction is obtained from the datasource which is likely shared across storages created by the same `MapStorageProvider` for different areas; here the identifier (and thus the transaction) would be shared within the session across areas.

**Comment by hmlnarik** [comment]
Attached is the diagram of the setup that we agreed on in the call:
- JPA transaction (which holds reference to DB transaction) is created lazily and enlisted in keycloak session upon first call to `getStorage`
- per-area transactions (e.g. [JpaGroupMapKeycloakTransaction](https://github.com/keycloak/keycloak/blob/6e181a51d52562887c2c25b5095799f31e115d30/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupMapKeycloakTransaction.java)) can for its operations rely on existence on JPA transaction, and thus the [transaction-processing methods](https://github.com/keycloak/keycloak/blob/6e181a51d52562887c2c25b5095799f31e115d30/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java#L166-L186) will be empty.
- A per-area map storage instance may contain a direct or indirect reference to the main JPA transaction, (e.g. by reference to an entity manager or a specific field referring directly the JPA transaction - as needed)

![Transaction processing in JPA transactions](https://user-images.githubusercontent.com/5391360/162762276-25cf79a2-cf62-4e6e-b1c3-0df535e701db.svg)

**Reply by sguilhen** [reply]
I think that, for this to work, the `JpaMapStorageProviderFactory` should keep the created `JpaMapStorageProvider` in the session and always return the same instance whenever `create` is called on it in the same session. The reason being that `JpaMapStorageProvider` is created with a new `EntityManager` and we would like to ensure that all area-specific transactions use the same entity manager, no? 

Otherwise something like this could happen:
 - factory creates `JpaMapStorageProvider` **A**. When `getStorage`is called, it checks the session, sees that no JPA transaction has been created yet, and creates one using its `EntityManager` instance. This provider returns a `MapStorage` that uses the same entity manager when instantiating new area-specific transactions so they can perform their operations using this entity manager.
 - now same factory creates `JpaMapStorageProvider` **B**. When `getStorage` is called, it checks the session and finds the JPA transaction, so it doesn't re-create or enlist it. Problem now is that this instance **B** will return a `MapStorage` that will use a different entity manager when creating new area-specific transactions. So, the area-specific tx will use a different `EntityManager` instance for its CRUD operations and this will cause it to not participate in the same transaction that was created by the instance **A**'s entity manager.

**Reply by hmlnarik** [reply]
You're correct. This is handled in https://github.com/keycloak/keycloak/blob/a6dd9dc0f1605ce0ac2b424df10e15a6eff6ff70/services/src/main/java/org/keycloak/services/DefaultKeycloakSession.java#L353-L370 but if there is a different way to obtain the `JpaMapStorageProvider` instance, then we should better ATM put the logic into the `create` method

**Reply by sguilhen** [reply]
The thing is currently we don't invoke this method when obtaining the map storage. When a `Map*ProviderFactory` creates a particular provider, it calls `getStorage`:

https://github.com/keycloak/keycloak/blob/ed79c2a86174daa3553f98af539063eb948a0ea1/model/map/src/main/java/org/keycloak/models/map/common/AbstractMapProviderFactory.java#L59-L65

Which essentially obtains the storage provider factory and then calls `create(session)` on it instead of calling the method you've referenced. This has the downside of the session not being able to track and cache the particular storage provider, so I wonder if we shouldn't change the `getStorage` implementation to call this method instead and reap the benefits of the session cache. Do you envision any problems with this change? For JPA we surely want the same provider instance to be used for the duration of the session and I can't see any of the other implementations requiring a new instance of their provider every time `getStorage` is called.

**Reply by hmlnarik** [reply]
Ah, yes, that's right. The reason is that in area `Map*Provider`s, there is a `storage` option in `config` which can either point to system-wide storage, adjusted storage config, or is just not there. It's related to #10324. In that case, let's go with the workaround you suggested (i.e. implementing the in-session caching of the created `JpaMapStorageProvider`s), just add a comment in appropriate place in the code that should be revisited once #10324 is resolved

**Reply by sguilhen** [reply]
I don't even think we will have to revisit this later because the `getComponentProvider` method uses a hash for key that includes the `componentId`, and this `componentId` will be different for each profile. So as long as the component id is computed in a way that takes the profile into consideration (which is something included in #10324 discussion) we should have proper providers being handed out by `session.getComponentProvider`.

**Reply by sguilhen** [reply]
This work is tracked by #11230
